code/modules/admin/smites/puzzle.dm

```diff
@@ -0,0 +1,330 @@
+	var/turf/center = get_turf(src)
+	switch(id)
+		if(1)
+			return get_step(center,NORTHWEST)
+		if(2)
+			return get_step(center,NORTH)
+		if(3)
+			return get_step(center,NORTHEAST)
+		if(4)
+			return get_step(center,WEST)
+		if(5)
+			return center
+		if(6)
+			return get_step(center,EAST)
+		if(7)
+			return get_step(center,SOUTHWEST)
+		if(8)
+			return get_step(center,SOUTH)
+		if(9)
+			return get_step(center,SOUTHEAST)
+
+/obj/effect/sliding_puzzle/Initialize(mapload)
+	..()
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/effect/sliding_puzzle/LateInitialize()
+	if(auto_setup)
+		setup()
+
+/obj/effect/sliding_puzzle/proc/check_setup_location()
+	for(var/id in 1 to 9)
+		var/turf/T = get_turf_for_id(id)
+		if(!T)
+			return FALSE
+		if(istype(T,/turf/closed))
+			return FALSE
+	return TRUE
+
+
+/obj/effect/sliding_puzzle/proc/validate()
+	if(finished)
+		return
+
+	if(elements.len < 8) //Someone broke it
+		qdel(src)
+
+	//Check if everything is in place
+	for(var/id in 1 to 9)
+		var/target_turf = get_turf_for_id(id)
+		var/obj/structure/puzzle_element/E = locate() in target_turf
+		if(id == empty_tile_id && !E) // This location should be empty.
+			continue
+		if(!E || E.id != id) //wrong tile or no tile at all
+			return
+	//Ding ding
+	finish()
+
+/obj/effect/sliding_puzzle/Destroy()
+	if(LAZYLEN(elements))
+		for(var/obj/structure/puzzle_element/E in elements)
+			E.source = null
+		elements.Cut()
+	return ..()
+
+#define COLLAPSE_DURATION 7
+
+/obj/effect/sliding_puzzle/proc/finish()
+	finished = TRUE
+	for(var/mob/M in range(7,src))
+		shake_camera(M, COLLAPSE_DURATION , 1)
+	for(var/obj/structure/puzzle_element/E in elements)
+		E.collapse()
+
+	dispense_reward()
+
+/obj/effect/sliding_puzzle/proc/dispense_reward()
+	new reward_type(get_turf(src))
+
+/obj/effect/sliding_puzzle/proc/is_solvable()
+	var/list/current_ordering = list()
+	for(var/obj/structure/puzzle_element/E in elements_in_order())
+		current_ordering += E.id
+
+	var/swap_tally = 0
+	for(var/i in 1 to current_ordering.len)
+		var/checked_value = current_ordering[i]
+		for(var/j in i to current_ordering.len)
+			if(current_ordering[j] < checked_value)
+				swap_tally++
+
+	return swap_tally % 2 == 0
+
+//swap two tiles in same row
+/obj/effect/sliding_puzzle/proc/make_solvable()
+	var/first_tile_id = 1
+	var/other_tile_id = 2
+	if(empty_tile_id == 1 || empty_tile_id == 2) //Can't swap with empty one so just grab some in second row
+		first_tile_id = 4
+		other_tile_id = 5
+
+	var/turf/T1 = get_turf_for_id(first_tile_id)
+	var/turf/T2 = get_turf_for_id(other_tile_id)
+
+	var/obj/structure/puzzle_element/E1 = locate() in T1
+	var/obj/structure/puzzle_element/E2 = locate() in T2
+
+	E1.forceMove(T2)
+	E2.forceMove(T1)
+
+/proc/cmp_xy_desc(atom/movable/A,atom/movable/B)
+	if(A.y > B.y)
+		return -1
+	if(A.y < B.y)
+		return 1
+	if(A.x > B.x)
+		return 1
+	if(A.x < B.x)
+		return -1
+	return 0
+
+/obj/effect/sliding_puzzle/proc/elements_in_order()
```

Comment:
this procs looks useless, just put sortTim(elements,cmp=/proc/cmp_xy_desc) at the correct place or make that a macro